mailpoet/lib/Automation/Integrations/Core/Filters/DateTimeFilter.php

```diff
@@ -0,0 +1,173 @@
+      return $this->matchesInTheLast($condition, $filterValue, $value);
+    }
+
+    // on the day(s) of the week
+    if ($condition === self::CONDITION_ON_THE_DAYS_OF_THE_WEEK) {
+      return $this->matchesOnTheDaysOfTheWeek($filterValue, $value);
+    }
+
+    // other conditions
+    if (!is_string($filterValue) || !$value instanceof DateTimeInterface) {
+      return false;
+    }
+
+    $datetime = $this->convertToWpTimezone($value);
+    switch ($condition) {
+      case 'before':
+        $ref = DateTimeImmutable::createFromFormat(self::FORMAT_DATETIME, $filterValue, $this->timezone);
+        return $ref && $datetime < $ref;
+      case 'after':
+        $ref = DateTimeImmutable::createFromFormat(self::FORMAT_DATETIME, $filterValue, $this->timezone);
+        return $ref && $datetime > $ref;
+      case 'on':
+        return $datetime->format(self::FORMAT_DATE) === $filterValue;
+      case 'not-on':
+        return $datetime->format(self::FORMAT_DATE) !== $filterValue;
+      default:
+        return false;
+    }
+  }
+
+  /** @param mixed $value */
+  private function matchesSet(string $condition, $value): bool {
+    switch ($condition) {
+      case self::CONDITION_IS_SET:
+        return $value !== null;
+      case self::CONDITION_IS_NOT_SET:
+        return $value === null;
+      default:
+        return false;
+    }
+  }
+
+  /**
+   * @param mixed $filterValue
+   * @param mixed $value
+   */
+  private function matchesInTheLast(string $condition, $filterValue, $value): bool {
+    if (!is_array($filterValue) || !isset($filterValue['number']) || !isset($filterValue['unit']) || !$value instanceof DateTimeInterface) {
+      return false;
+    }
+
+    $number = $filterValue['number'];
+    $unit = $filterValue['unit'];
+    if (!is_integer($number) || !in_array($unit, ['days', 'weeks', 'months'], true)) {
+      return false;
+    }
+
+    $now = new DateTimeImmutable('now', $this->timezone);
+    $ref = $now->modify("-$number $unit");
+    $matches = $ref <= $value && $value <= $now;
+    return $condition === self::CONDITION_IN_THE_LAST ? $matches : !$matches;
+  }
+
+  /**
+   * @param mixed $filterValue
+   * @param mixed $value
+   */
+  private function matchesOnTheDaysOfTheWeek($filterValue, $value): bool {
+    if (!is_array($filterValue) || !$value instanceof DateTimeInterface) {
+      return false;
+    }
+    foreach ($filterValue as $day) {
+      if (!is_integer($day) || $day < 0 || $day > 6) {
+        return false;
+      }
+    }
+
+    $date = $this->convertToWpTimezone($value);
+    $day = (int)$date->format('w');
+    return in_array($day, $filterValue, true);
+  }
+
+  private function convertToWpTimezone(DateTimeInterface $datetime): DateTimeImmutable {
+    $value = DateTimeImmutable::createFromFormat('U', (string)$datetime->getTimestamp(), $this->timezone);
+    if (!$value) {
+      throw new InvalidStateException('Failed to convert datetime to WP timezone');
+    }
+    return $value;
+  }
```

Comment:
Hmm... If we call it `convertToWpTimezone` and bind it to the WpTimezone, I would at least name the `private $timezone` `private $wpTimezone` I think.

I like that it's a `\DateTimeZone` in the constructor and not the `WordPress`-wrapper, but it might make sense to use `$this->wp->wpTimezone()` instead here, if you want to tie it to `WpTimezone`.

Another thought would be `convertToLocalTimezone()` maybe, then I think `private $localeTimezone` would be appropriate :thinking:

Reply:
Great point! Renamed to `localTimezone`.